source/mesh_deformation/function.cc

```diff
@@ -0,0 +1,124 @@
+
+#include <deal.II/numerics/vector_tools.h>
+
+namespace aspect
+{
+  namespace MeshDeformation
+  {
+    template <int dim>
+    Function<dim>::Function()
+      :
+      function(dim)
+    {}
+
+
+
+    template <int dim>
+    void
+    Function<dim>::update ()
+    {
+      function.set_time(this->get_time());
+    }
+
+
+
+    template <int dim>
+    void
+    Function<dim>::deformation_constraints(const DoFHandler<dim> &free_surface_dof_handler,
+                                           ConstraintMatrix &mesh_constraints) const
+    {
+      for (std::set<types::boundary_id>::const_iterator p = this->get_parameters().free_surface_boundary_indicators.begin();
+           p != this->get_parameters().free_surface_boundary_indicators.end(); ++p)
+        {
+          VectorTools::interpolate_boundary_values (free_surface_dof_handler,
+                                                    *p,
+                                                    function,
+                                                    mesh_constraints);
+        }
+    }
+
+
+
+    template <int dim>
+    void Function<dim>::declare_parameters(ParameterHandler &prm)
+    {
+      prm.enter_subsection ("Free surface");
+      {
+        prm.enter_subsection ("Function");
+        {
+          Functions::ParsedFunction<dim>::declare_parameters (prm, dim);
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection ();
+    }
+
+    template <int dim>
+    void Function<dim>::parse_parameters(ParameterHandler &prm)
+    {
+      prm.enter_subsection ("Free surface");
+      {
+        prm.enter_subsection("Function");
+        try
+          {
+            function.parse_parameters (prm);
+          }
+        catch (...)
+          {
+            std::cerr << "ERROR: FunctionParser failed to parse\n"
+                      << "\t'Mesh deformation.Function'\n"
+                      << "with expression\n"
+                      << "\t'" << prm.get("Function expression") << "'"
+                      << "More information about the cause of the parse error \n"
+                      << "is shown below.\n";
+            throw;
+          }
+        prm.leave_subsection();
+
+      }
+      prm.leave_subsection ();
+    }
+  }
+}
+
+
+// explicit instantiation of the functions we implement in this file
+namespace aspect
+{
+  namespace MeshDeformation
+  {
+    ASPECT_REGISTER_MESH_DEFORMATION_MODEL(Function,
+                                           "function",
+                                           "A plugin, which prescribes the surface mesh to "
+                                           "deform according to an analytically prescribed "
+                                           "function. Note that the function prescribes a "
+                                           "deformation velocity, i.e. the return value of "
+                                           "the function is multiplied by the time step length "
```

Comment:
The way I understand it, the return value of the function is not multiplied by the time step length here, but the velocity solution computed with the constraints provided here is multiplied by the time step length to compute the mesh displacement. The way it is formulated here, it seems this plugin returns the displacement instead of the displacement velocity.